src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadonlyStructTests.cs

```diff
@@ -0,0 +1,522 @@
+}");
+        }
+
+        [Fact]
+        public void InvokeOnThis()
+        {
+            var text = @"
+class Program
+{
+    static void Main()
+    {
+        default(S1).Test();
+    }
+    readonly struct S1
+    {
+        public void Test()
+        {
+            System.Console.Write(this.M1());
+            System.Console.Write(ToString());
+        }
+        public string M1()
+        {
+            return ""1"";
+        }
+        public override string ToString()
+        {
+            return ""2"";
+        }
+    }
+}
+";
+
+            var comp = CompileAndVerify(text, new[] { ValueTupleRef, SystemRuntimeFacadeRef }, parseOptions: TestOptions.Regular, verify: false, expectedOutput: @"12");
+
+            comp.VerifyIL("Program.S1.Test()", @"
+{
+  // Code size       29 (0x1d)
+  .maxstack  1
+  IL_0000:  ldarg.0
+  IL_0001:  call       ""string Program.S1.M1()""
+  IL_0006:  call       ""void System.Console.Write(string)""
+  IL_000b:  ldarg.0
+  IL_000c:  constrained. ""Program.S1""
+  IL_0012:  callvirt   ""string object.ToString()""
+  IL_0017:  call       ""void System.Console.Write(string)""
+  IL_001c:  ret
+}");
+
+            comp.VerifyIL("Program.Main()", @"
+{
+  // Code size       18 (0x12)
+  .maxstack  1
+  .locals init (Program.S1 V_0)
+  IL_0000:  ldloca.s   V_0
+  IL_0002:  initobj    ""Program.S1""
+  IL_0008:  ldloc.0
+  IL_0009:  stloc.0
+  IL_000a:  ldloca.s   V_0
+  IL_000c:  call       ""void Program.S1.Test()""
+  IL_0011:  ret
+}");
+
+        }
+
+        [Fact]
+        public void InvokeOnThisBaseMethods()
+        {
+            var text = @"
+class Program
+{
+    static void Main()
+    {
+        default(S1).Test();
+    }
+    readonly struct S1
+    {
+        public void Test()
+        {
+            System.Console.Write(this.GetType());
+            System.Console.Write(ToString());
+        }
+    }
+}
+";
+
+            var comp = CompileAndVerify(text, new[] { ValueTupleRef, SystemRuntimeFacadeRef }, parseOptions: TestOptions.Regular, verify: false, expectedOutput: @"Program+S1Program+S1");
+
+            comp.VerifyIL("Program.S1.Test()", @"
+{
+  // Code size       39 (0x27)
+  .maxstack  1
+  IL_0000:  ldarg.0
+  IL_0001:  ldobj      ""Program.S1""
+  IL_0006:  box        ""Program.S1""
+  IL_000b:  call       ""System.Type object.GetType()""
+  IL_0010:  call       ""void System.Console.Write(object)""
+  IL_0015:  ldarg.0
+  IL_0016:  constrained. ""Program.S1""
+  IL_001c:  callvirt   ""string object.ToString()""
+  IL_0021:  call       ""void System.Console.Write(string)""
+  IL_0026:  ret
+}");
+        }
+
+        [Fact]
+        public void AssignThis()
+        {
+            var text = @"
+class Program
+{
+    static void Main()
+    {
+    }
+
+    static void TakesRef(ref S1 arg){}
+    static void TakesRef(ref int arg){}
+
+    readonly struct S1
+    {
+        readonly int x;
+
+        public S1(int i)
+        {            
+            x = i; // OK
```

Comment:
>x = i; // OK [](start = 12, length = 12)

I think we should have a test that runs the code to verify runtime behavior and probably IL as well.